databuilder/extractor/kafka_source_extractor.py

```diff
@@ -0,0 +1,144 @@
+
+    It uses the "micro-batch" concept to ingest data from a given Kafka topic and
+    persist into downstream sink.
+    Once the publisher commit successfully, it will trigger the extractor's callback to commit the
+    consumer offset.
+    """
+    # The dict of Kafka consumer config
+    CONSUMER_CONFIG = 'consumer_config'
+    # The consumer group id. Ideally each Kafka extractor should only associate with one consumer group.
+    CONSUMER_GROUP_ID = 'group.id'
+    # We don't deserde the key of the message.
+    # CONSUMER_VALUE_DESERDE = 'value.deserializer'
+    # Each Kafka extractor should only consume one single topic. We could extend to consume more topic if needed.
+    TOPIC_NAME_LIST = 'topic_name_list'
+
+    # Time out config. It will abort from reading the Kafka topic after timeout is reached. Unit is seconds
+    CONSUMER_TOTAL_TIMEOUT = 'consumer_total_timeout'
+
+    # The timeout for consumer polling messages. Default to 1 sec
+    CONSUMER_POLL_TIMEOUT = 'consumer_poll_timeout'
+
+    RAW_VALUE_TRANSFORMER = 'raw_value_transformer'
+
+    def init(self, conf):
+        # type: (ConfigTree) -> None
+        self.conf = conf
+        # pyhocon doesn't allow dot to be in value dict,group.id -> group_id
+        self.consumer_config = conf.get_config(KafkaSourceExtractor.CONSUMER_CONFIG).\
+            as_plain_ordered_dict()
+
+        # convert group_id back to group.id
+        # noted: underscore is never used in Kafka consumer config
+        self.consumer_config = {k.replace('_', '.'): v for k, v in self.consumer_config.items()}
+
+        self.topic_names = conf.get_list(KafkaSourceExtractor.TOPIC_NAME_LIST)  # type: list
+
+        if not self.topic_names:
+            raise Exception('Kafka topic needs to be provided by the user.')
+
+        self.consumer_total_timeout = conf.get_int(KafkaSourceExtractor.CONSUMER_TOTAL_TIMEOUT,
+                                                   default=10)
+
+        self.consumer_poll_timeout = conf.get_int(KafkaSourceExtractor.CONSUMER_POLL_TIMEOUT,
+                                                  default=1)
+
+        # Transform the protoBuf message with a transformer
+        self.val_transformer = conf.get(KafkaSourceExtractor.RAW_VALUE_TRANSFORMER)
+        if self.val_transformer is None:
+            raise Exception('A message transformer should be provided.')
+
+        # Consumer init
+        try:
+            self.consumer = Consumer(self.consumer_config)
+            # TODO: to support only consume a subset of partitions.
+            self.consumer.subscribe(self.topic_names)
+        except Exception:
+            raise RuntimeError('Consumer could not start correctly!')
+
+    def extract(self):
+        # type: () -> Any
+        """
+        :return: Provides a record or None if no more to extract
+        """
+        records = self.consume()
+        for record in records:
+            try:
+                transform_record = self.val_transformer(record)
```

Comment:
this part needs to update:
1. verify whether val_transformer is a transformer class.
2. use val_transformer.transform(record)